Db-Services refactoring

TanVD opened the pull request:
1.Add exceptions in all db-services (#129917889)
2.Add unit tests in all db-services (#127461159, #126217043, #129917843)

jzuken commented on dashboard-service/src/main/java/com/qreal/wmp/dashboard/database/exceptions/Aborted.java:
```diff
@@ -0,0 +1,29 @@
+package com.qreal.wmp.dashboard.database.exceptions;
+
+public class Aborted extends Exception {
```
javadoc?

jzuken commented on dashboard-service/src/main/java/com/qreal/wmp/dashboard/database/exceptions/Aborted.java:
```diff
@@ -0,0 +1,29 @@
+        super(message, cause);
+        this.textCause = textCause;
+        this.fullClassName = fullClassName;
+
```
unnecessary empty line

jzuken commented on dashboard-service/src/main/java/com/qreal/wmp/dashboard/database/exceptions/ErrorConnection.java:
```diff
@@ -0,0 +1,20 @@
+package com.qreal.wmp.dashboard.database.exceptions;
+
+public class ErrorConnection extends Exception {
```
javadoc

jzuken commented on dashboard-service/src/main/java/com/qreal/wmp/dashboard/database/robots/client/RobotServiceImpl.java:
```diff
@@ -49,95 +57,169 @@ public void start() {
-                    "robot = {}", robot.getName(), e);
+            try {
+                idRobot = client.registerRobot(robot.toTRobot());
+            } catch (TIdAlreadyDefined e) {
```
we'd better crash with such coding errors in development mode and show corresponding error in diagram service

jzuken commented on thrift-definitions/exception/DbExceptions.thrift:
```diff
@@ -0,0 +1,34 @@
+    2: string message
+}
+
+//Used as a coding rule
```
why not unify the comments in all thrift declaration files?

jzuken commented on thrift-definitions/diagram-database/DiagramDbService.thrift:
```diff
@@ -2,12 +2,57 @@ namespace java com.qreal.wmp.thrift.gen
-    void deleteFolder(1: i64 id);
-    Diagram.TFolder getFolderTree(1: string username);
+    /**
+    * In case of exception save will be safely aborted
```
save operation will be safely aborted

jzuken commented on thrift-definitions/diagram-database/DiagramDbService.thrift:
```diff
@@ -2,12 +2,57 @@ namespace java com.qreal.wmp.thrift.gen
+    /**
+    * In case of exception save will be safely aborted
+    *
+    * TAborted will be thrown in following cases:
```
in the following cases

jzuken replied:
here and everywhere below

jzuken commented on thrift-definitions/diagram-database/DiagramDbService.thrift:
```diff
@@ -2,12 +2,57 @@ namespace java com.qreal.wmp.thrift.gen
+    * In case of exception save will be safely aborted
+    *
+    * TAborted will be thrown in following cases:
+    * 1. Folder to save diagram into doesn't exist
```
The folder to save the diagram into doesn't exist

jzuken commented on thrift-definitions/diagram-database/DiagramDbService.thrift:
```diff
@@ -2,12 +2,57 @@ namespace java com.qreal.wmp.thrift.gen
+    i64 saveDiagram(1: Diagram.TDiagram diagram) throws(1: DbExceptions.TIdAlreadyDefined alreadyDefined,
+                                                        2: DbExceptions.TAborted aborted);
+
+    //If not found diagram with specified id TNotFound will be thrown
```
TNotFound is thrown if the diagram with specified id is not found

jzuken commented on thrift-definitions/diagram-database/DiagramDbService.thrift:
```diff
@@ -2,12 +2,57 @@ namespace java com.qreal.wmp.thrift.gen
+
+
+    /**
+    * In case of exception delete will be safely aborted
```
delete operation will be safely aborted

jzuken commented on thrift-definitions/exception/DbExceptions.thrift:
```diff
@@ -0,0 +1,34 @@
+namespace java com.qreal.wmp.thrift.gen
+
+//Used in normal cases of excution (instead of returning null)
```
is this exception thrown always instead of returning null?

TanVD replied:
In search operations -- yes

jzuken commented on thrift-definitions/exception/DbExceptions.thrift:
```diff
@@ -0,0 +1,34 @@
+    2: string message
+}
+
+//Used in exceptional situations to informate other services and client
```
to inform

jzuken commented on thrift-definitions/exception/DbExceptions.thrift:
```diff
@@ -0,0 +1,34 @@
+    3: string fullClassName
+}
+
+//Used in situation of connection errors.
```
in case of

jzuken commented on thrift-definitions/robot-store-database/RobotDbService.thrift:
```diff
@@ -2,10 +2,55 @@ namespace java com.qreal.wmp.thrift.gen
+* RobotDbService uses client to UsersDbService to resolve
+* foreign key dependencies.
+*
+* Robot foreign key dependency of owner to Users table NOT checked,
```
I can't really parse this comment :)

jzuken commented on db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/dao/DiagramDaoImpl.java:
```diff
@@ -113,8 +114,7 @@ public boolean isExistsDiagram(Long diagramId) {
     @Override
-    public void rewriteDiagram(@NotNull Diagram diagram) throws Aborted {
-        logger.trace("rewriteDiagram method called with parameters: diagram = {}", diagram.getName());
+    public void rewriteDiagram(@NotNull Diagram diagram) throws Aborted {logger.trace("rewriteDiagram method called with parameters: diagram = {}", diagram.getName());
```
newline is lost somehow here

jzuken commented on db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/server/DiagramDbServiceHandler.java:
```diff
@@ -100,4 +101,19 @@ public TFolder getFolderTree(String username) throws TNotFound {
     }
+
+    /** Only for sake of testing. */
+    public DiagramDao getDiagramDao() {
```
do they really need to be public in the sake of testing?

TanVD replied:
I don't see any other option (excepting moving tests for server to server package)

jzuken commented on db-services/db-diagram-service/src/test/java/com/qreal/wmp/db/diagram/test/dao/DaoDiagramComplexTest.java:
```diff
@@ -0,0 +1,70 @@
+        node.setType("type");
+
+        Property propertyNode = new Property();
+        propertyNode.setName("namePropertyNode");
```
why don't we create some constructors to eliminate these setters calls?

jzuken commented on db-services/db-diagram-service/src/test/java/com/qreal/wmp/db/diagram/test/dao/DaoDiagramComplexTest.java:
```diff
@@ -0,0 +1,70 @@
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+import org.springframework.transaction.annotation.Transactional;
+
+import static org.assertj.core.api.Assertions.assertThat;
```
I like Hamcrest assertions better :)

jzuken commented on db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/exceptions/Aborted.java:
```diff
@@ -1,9 +1,12 @@
 package com.qreal.wmp.db.diagram.exceptions;
 
+/** Aborted exception inform about inconsistency of DB or any other situation lead to abort of operation.*/
 public class Aborted extends Exception {
```
I guess exceptions' names should end with 'Exception', like AbortedException

jzuken commented on db-services/db-diagram-service/src/test/java/com/qreal/wmp/db/diagram/test/server/DbServiceHandlerDiagramTest.java:
```diff
@@ -0,0 +1,180 @@
+@Transactional
+public class DbServiceHandlerDiagramTest {
+
+    public DiagramDbServiceHandler handler;
```
why public?

jzuken commented on db-services/db-diagram-service/src/test/java/com/qreal/wmp/db/diagram/test/dao/DaoFolderTest.java:
```diff
@@ -0,0 +1,141 @@
+@Transactional
+public class DaoFolderTest {
+    @Autowired
+    public DiagramDao diagramDao;
```
and why  dao fields are public?

jzuken commented on db-services/db-diagram-service/src/test/java/com/qreal/wmp/db/diagram/test/server/DbServiceHandlerDiagramTest.java:
```diff
@@ -0,0 +1,180 @@
+    /** Test saveDiagram operation for diagram. */
+    @Test
+    @Rollback
+    public void saveDiagram_correctInput_diagramDaoCalled() throws TAborted, TIdAlreadyDefined, Aborted {
```
it is common practice to just use "throws Exception" in test methods

jzuken commented on dashboard-service/src/main/java/com/qreal/wmp/dashboard/controller/MainController.java:
```diff
@@ -21,14 +23,25 @@
         logger.info("User {} requested main page.", AuthenticatedUser.getUserName());
 
-        User user = userService.findByUserName(AuthenticatedUser.getUserName());
+        User user = null;
```
we do model.addObject("robots", user.getRobots()); several lines below, I guess we sould handle this potential null value there

TanVD replied:
i'll handle it, but in normal situation user cannot be null, cause to call this method user should be authorized

jzuken commented on dashboard-service/src/main/java/com/qreal/wmp/dashboard/database/users/client/UserDetailsServiceImpl.java:
```diff
@@ -19,20 +23,30 @@
 
-        List<GrantedAuthority> authorities =
-                buildUserAuthority(user.getRoles());
+        List<GrantedAuthority> authorities = buildUserAuthority(user.getRoles());
```
and again, if the user is null?

TanVD replied:
If user was not found, NotFound will be thrown. But in case of ErrorConnection user can be null, i'll fix it.

jzuken committed:
Fix docs for db-robot-store-service

More grammar-nazi fixes

jzuken committed:
Update comments everywhere.

Tests are failing though. Need to create custom Spring profile for DAO tests in db services.

TanVD committed:
Merge branch 'master' into db-refactoring

TanVD committed:
Add AOP exception translatio for db-services.

Add exception translation from Thrift exceptions to domain-specific exceptions
using AOP. We use AspectJ around advice to catch TExceptions in
db-services clients and translate this exceptions to domain-specific.

jzuken commented on db-services/db-robot-store-service/src/main/java/com/qreal/wmp/db/robot/exceptions/translation/ExceptionTranslation.java:
```diff
@@ -0,0 +1,48 @@
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Component;
+
+@Aspect
```
I guess it would be better to add a comment here describing all this aspect stuff.

TanVD replied:
Ok, i'll add javadoc

jzuken commented on db-services/db-user-service/src/main/java/com/qreal/wmp/db/user/dao/UserDaoImpl.java:
```diff
@@ -43,114 +43,127 @@ public UserDaoImpl(SessionFactory sessionFactory, RobotService robotService, Dia
+                try {
+                    idRobot = robotService.register(robot);
+                } catch (TException e) {
+                    e.printStackTrace();
```
just print stack trace?

TanVD replied:
Yep, this exception should be translated by ExceptionTranslation aspect and should never be caught here. I'll think about hack solution.

TanVD committed:
Delete getters-setters. Add javadoc.

Getters and setters delete from model classes cause lombok generates them.
Also javadoc for ExceptionTranslation class add. And more minor fixes.

TanVD committed:
Add exception translation to first-line services.

Add exception translation to editor-service and dashboard service. Exception
translation performed by Aspect-J and Spring (using AOP techniques). Also
fix lombok bug with circular dependencies. Fix a lot of minor bugs and
style violations.

TanVD committed:
Codacy style fix

Fix e.printStackTrace() method calls and redundunt imports.
